### PR TITLE
GEODE-7577: Log when waiting for missing persistent members

### PR DESCRIPTION
**Pull request by kirklund**
GEODE-4267: Fix PersistentRecoveryOrderOldConfigDUnitTest

This started out trying to fix flaky failures in PersistentRecoveryOrderOldConfigDUnitTest and PersistentRecoveryOrderDUnitTest (GEODE-4267). I found that the log statement about waiting for missing persistent members was no longer working, so I ended up filing GEODE-7577 and fixing that as well.

New tests:
* MissingDiskStoreAcceptanceTest -- acceptance test to verify that a server logs when it is waiting for missing persistent members
* MembershipChangeListenerFactoryTest -- unit test for the new MembershipChangeListenerFactory
* New test methods in PersistenceAdvisorImplTest for the formatting of the messages

Changes:
* Separate StartupStatusListenerRegistry from StartupStatus to allow non-static injection of StartupStatus into classes that use it
* Add CollectionTransformer interface to TransformUtils to facilitate unit testing of classes depending on it
* Add new MembershipChangeListenerFactory with validation to ensure that warningDelay and pollDuration have correct values in relation to each other (to prevent recurrences of GEODE-7577)
* Fixed the formatting of missing persistent members messages in PersistenceAdvisorImpl
* Fixed warnings and minor issues primarily for readability and testability in every class touched
* Added 'boolean shouldSyncForCrashedMember(InternalDistributedMember id)' to InternalRegion and pulled the javadocs up to the interface
* Major cleanup of PersistentRecoveryOrderDUnitTest including replace usage of Admin API with usage of DistributedSystemMXBean
* Replace Region close with Cache close in PersistentRecoveryOrderDUnitTest test that needs to list and revoke missing disk-stores

Fixes to PersistentRecoveryOrderDUnitTest:
* Replace thread sleeps with CountDownLatch
* Add a couple await calls
* Actual fix involved preventing late-running asynchronous actions that resulted in logging DistributedSystemDisconnectedException AFTER the test already removed the IgnoredException

The changes to the formatting of the messages about missing persistent members restores the old format that was correct. The formatting was broken when LocalizedStrings was removed.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/process/StartupStatusListenerRegistry.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * the License.
+ */
+package org.apache.geode.internal.process;
+
+import org.apache.geode.annotations.internal.MakeNotStatic;
+
+public class StartupStatusListenerRegistry {
+
+  @MakeNotStatic
+  private static volatile StartupStatusListener listener;
```
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/internal/process/StartupStatusListenerRegistry.java]: Change to AtomicReference

File: geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java [hunk 2]
```diff
@@ -1382,9 +1346,46 @@ private void basicRemoveIndex(int index) {
     }
   }
 
+  // ------------------------------------ inner classes --------------------------------------------
```
Comment (igodwin) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: Remove comment.

File: geode-assembly/src/acceptanceTest/java/org/apache/geode/cache/persistence/MissingDiskStoreAcceptanceTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    String createRegionCommand = String.join(" ",
+        "create region",
+        "--name=" + REGION_NAME,
+        "--type=REPLICATE_PERSISTENT");
+
+    gfshRule.execute(startLocatorCommand, startServer1Command, startServer2Command,
+        createRegionCommand);
+
+    clientCache = (InternalClientCache) new ClientCacheFactory()
```
Comment (upthewaterspout) [inline geode-assembly/src/acceptanceTest/java/org/apache/geode/cache/persistence/MissingDiskStoreAcceptanceTest.java]: Why is this casting to InternalClientCache?

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/persistence/PersistentRecoveryOrderDUnitTest.java [hunk 4]
```diff
@@ -131,107 +187,125 @@ public Properties getDistributedSystemProperties() {
...
-
-    createPersistentRegion(vm0);
-    createPersistentRegion(vm1);
+    for (VM vm : toArray(vm0, vm1)) {
+      vm.invoke(() -> createReplicateRegion(regionName, getDiskDirs(getVMId())));
+    }
 
-    putAnEntry(vm0);
-    closeRegion(vm0);
+    vm0.invoke(() -> {
```
Comment (upthewaterspout) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/persistence/PersistentRecoveryOrderDUnitTest.java]: This seems a little less readable after you inlined these utility methods.

File: geode-core/src/main/java/org/apache/geode/internal/cache/persistence/MembershipChangeListenerFactory.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import static java.util.Objects.requireNonNull;
+import static org.apache.geode.internal.lang.SystemPropertyHelper.PERSISTENT_VIEW_RETRY_TIMEOUT_SECONDS;
+import static org.apache.geode.internal.lang.SystemPropertyHelper.getProductIntegerProperty;
+
+import java.time.Duration;
+import java.util.function.BooleanSupplier;
+
+import org.apache.geode.CancelCriterion;
+
+public class MembershipChangeListenerFactory {
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/internal/cache/persistence/MembershipChangeListenerFactory.java]: Be nice to have some javadocs describing what this class does.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java (hunk 2 above)
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java]: Deleted

File: geode-assembly/src/acceptanceTest/java/org/apache/geode/cache/persistence/MissingDiskStoreAcceptanceTest.java (hunk 3 above)
Comment (kirklund) [reply geode-assembly/src/acceptanceTest/java/org/apache/geode/cache/persistence/MissingDiskStoreAcceptanceTest.java]: Removed. It just uses ClientCache so I'm not sure why the field was InternalClientCache.

File: geode-core/src/main/java/org/apache/geode/internal/cache/persistence/MembershipChangeListenerFactory.java (hunk 5 above)
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/internal/cache/persistence/MembershipChangeListenerFactory.java]: Added

**Pull request merged**